config/v1/types_infrastructure.go

```diff
@@ -840,14 +840,14 @@ type BareMetalPlatformSpec struct {
 	//
 	// +kubebuilder:validation:MaxItems=2
 	// +kubebuilder:validation:XValidation:rule="size(self) == 2 && isIP(self[0]) && isIP(self[1]) ? ip(self[0]).family() != ip(self[1]).family() : true",message="ingressIPs must contain at most one IPv4 address and at most one IPv6 address"
-	// +listType=set
+	// +listType=atomic
 	// +optional
 	IngressIPs []IP `json:"ingressIPs"`
 
 	// machineNetworks are IP networks used to connect all the OpenShift cluster
 	// nodes. Each network is provided in the CIDR format and should be IPv4 or IPv6,
 	// for example "10.0.0.0/8" or "fd00::/8".
-	// +listType=set
+	// +listType=atomic
 	// +kubebuilder:validation:MaxItems=32
 	// +optional
 	MachineNetworks []CIDR `json:"machineNetworks"`
```

Comment:
Need CEL to keep set semantic here and on every other MachineNetworks

Reply:
Right, adding
```
// +kubebuilder:validation:XValidation:rule="self.all(x, self.exists_one(y, x == y))"
```
and testing. Can you confirm it's okay to have CEL only in `Spec` and leave it out of `Status` as only the former is supposed to accept user input?

Reply:
Prefer to have it in both, because while only spec is supposed to be accessible by user, they can still edit status

Reply:
Fair enough. Given that API and Ingress VIP have CELs only in Spec, do you want me to also add them to Status in this PR ?

Reply:
If we are confident that isn't going to break anything, then yes.

The no breaking way to add it is to add ratcheting, prefix with `self == oldSelf ||`

Reply:
Does CNO do any validation that the two IP addresses in status are separate families?

Reply:
>If we are confident that isn't going to break anything, then yes.

I think "confident" is a stretch because in the past this field was basically copy-pasted from install-config.yaml and all the validations were happening in o/installer. I believe it wouldn't break anything, but it's more faith than certainty

>Does CNO do any validation that the two IP addresses in status are separate families?

Yes, we have https://github.com/openshift/cluster-network-operator/blob/master/pkg/controller/infrastructureconfig/validations.go#L40-L50 that is validating that if you provide more than 1 VIP, you need to have multiple IP families. So we will not allow setting 2 IPs that are IPv4-only or IPv6-only

Reply:
> I think "confident" is a stretch because in the past this field was basically copy-pasted from install-config.yaml and all the validations were happening in o/installer. I believe it wouldn't break anything, but it's more faith than certainty

Was the installer doing any validation?

> Yes, we have https://github.com/openshift/cluster-network-operator/blob/master/pkg/controller/infrastructureconfig/validations.go#L40-L50 that is validating that if you provide more than 1 VIP, you need to have multiple IP families. So we will not allow setting 2 IPs that are IPv4-only or IPv6-only

At what stage does this occur? Does it cause the cluster to degrade or anything?

Reply:
Ok, and when were these APIs introduced as GA? I think we are safe to copy the validation to the status provided the installer has already validated this and you go degraded if the status gets edited.

If the status gets edited to invalid, does it get put back by anything?

Reply:
>when were these APIs introduced as GA

https://github.com/openshift/api/pull/1152 (merged in August 2022) + https://github.com/openshift/enhancements/pull/1048 (looks like 4.11)

>If the status gets edited to invalid, does it get put back by anything?

Yes. In the master branch of CNO if you modify Status without modifying Spec, Status will get reverted because operator reconciles Status to match Spec. So for a scenario

1) start from valid Spec and Status
2) update with invalid Status (without modifying Spec)

you end up with no changes applied because Spec is authoritative over Status. You would need to modify Spec *and* Status, but then Spec is validated via CEL already now and that will stop you.

Hope this explanation is enough to support statement "we have no way of introducing illegal values into Status"